libass/ass_font.c

```diff
@@ -26,15 +26,165 @@
 #include FT_GLYPH_H
 #include FT_TRUETYPE_TABLES_H
 #include FT_OUTLINE_H
+#include FT_TRUETYPE_IDS_H
 #include <limits.h>
 
+/**
+ * We don't exclude Cygwin for Windows since we use WideCharToMultiByte only,
+ * this shall not violate any Cygwin restrictions on Windows APIs.
+ */
+#if defined(_WIN32)
+#define WIN32_LEAN_AND_MEAN
+#include <windows.h>
+#elif defined(CONFIG_ICONV)
+#include <iconv.h>
+#endif
+
 #include "ass.h"
 #include "ass_library.h"
 #include "ass_font.h"
 #include "ass_fontselect.h"
 #include "ass_utils.h"
 #include "ass_shaper.h"
 
+
+/**
+ * Convert a UCS-4 code unit to a packed uint32_t in given multibyte encoding
+ */
+#if defined(_WIN32)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    // map freetype encoding to windows codepage
+    UINT codepage;
+    switch (encoding) {
+    case FT_ENCODING_MS_SJIS:
+        codepage = 932;
+        break;
+    case FT_ENCODING_MS_GB2312:
+        codepage = 936;
+        break;
+    case FT_ENCODING_MS_BIG5:
+        codepage = 950;
+        break;
+    case FT_ENCODING_MS_WANSUNG:
+        codepage = 949;
+        break;
+    case FT_ENCODING_MS_JOHAB:
+        codepage = 1361;
+        break;
+    default:
+        return 0;
+    }
+
+    wchar_t input_buffer[2];
+    size_t inbuf_size;
+
+    // encode surrogate pair, assuming codepoint > 0 && codepoint <= 10FFFF
+    if (codepoint >= 0x10000) {
+        // surrogate pair required
+        inbuf_size = 2;
+        uint32_t code = (codepoint - 0x10000);
+        input_buffer[0] = 0xD800 | (code >> 10);
+        input_buffer[1] = 0xDC00 | (code & 0x3FF);
+    } else {
+        inbuf_size = 1;
+        input_buffer[0] = codepoint;
+    }
+
+    // do convert
+    char output_buffer[4];
+    BOOL conversion_fail;
+    int outbuf_size = WideCharToMultiByte(codepage, WC_NO_BEST_FIT_CHARS, input_buffer, inbuf_size,
+                                          output_buffer, sizeof(output_buffer), NULL, &conversion_fail);
+    if (outbuf_size == 0 || conversion_fail)
+        return 0;
+
+    uint32_t result = 0;
+    for (size_t i = 0; i < outbuf_size; ++i) {
+        result <<= 8;
+        result |= (uint8_t) output_buffer[i];
+    }
+
+    return result;
+}
+#elif defined(CONFIG_ICONV)
+static uint32_t convert_unicode_to_mb(FT_Encoding encoding, uint32_t codepoint)
+{
+    typedef struct { const char* names[8]; } EncodingList;
```

Comment:
Another style nit, here and below: libass prefers to put the space before the asterisks, like so:

```c
const char *names[8]
const char **encoding_str
char *inbuf
char *outbuf
```